src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs

```diff
@@ -290,11 +290,12 @@ static void Goo(this string x)
         Console.WriteLine(x);
     }
 }";
+            // ILVerify: Unrecognized arguments for delegate .ctor. { Offset = 21 }
```

Comment:
>ILVerify: Unrecognized arguments for delegate .ctor. { Offset = 21 } [](http://example.com/codeflow?start=15&length=68)

Is this error reported for a delegate created from an extension method? And does it indicate a bug in Roslyn or in ILVerify?

Reply:
Since PEVerify passes, I think it's an ILVerify limitation.
The IL looks fine as far as I can tell:
```
.class private auto ansi abstract sealed beforefieldinit Program
	extends [netstandard]System.Object
{
	.custom instance void [netstandard]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = (
		01 00 00 00
	)
	// Methods
	.method private hidebysig static 
		void Main () cil managed 
	{
		// Method begins at RVA 0x2050
		// Code size 70 (0x46)
		.maxstack 2
		.entrypoint
		IL_0000: ldstr "ABC"
		IL_0005: call void Program::Goo(string)
		IL_000a: ldstr "123"
		IL_000f: ldftn void Program::Goo(string)
		IL_0015: newobj instance void [netstandard]System.Action::.ctor(object, native int)
		IL_001a: dup
		IL_001b: callvirt instance void [netstandard]System.Action::Invoke()
		IL_0020: ldftn instance void [netstandard]System.Action::Invoke()
		IL_0026: newobj instance void [netstandard]System.Action::.ctor(object, native int)
		IL_002b: callvirt instance void [netstandard]System.Action::Invoke()
		IL_0030: ldstr "xyz"
		IL_0035: ldftn void Program::Goo(string)
		IL_003b: newobj instance void [netstandard]System.Action::.ctor(object, native int)
		IL_0040: callvirt instance void [netstandard]System.Action::Invoke()
		IL_0045: ret
	} // end of method Program::Main
	.method private hidebysig static 
		void Goo (
			string x
		) cil managed 
	{
		.custom instance void [netstandard]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = (
			01 00 00 00
		)
		// Method begins at RVA 0x20a2
		// Code size 7 (0x7)
		.maxstack 8
		IL_0000: ldarg.0
		IL_0001: call void [netstandard]System.Console::WriteLine(string)
		IL_0006: ret
	} // end of method Program::Goo
} // end of class Program
```

Reply:
Confirmed the issue.
ILVerify check whether `Program.Goo(string)` can be assigned to delegate type `Action` and fails. It doesn't account for `this`/extension scenario.

```
        void CheckIsDelegateAssignable(MethodDesc ftn, TypeDesc delegateType)
        {
            if (!IsDelegateAssignable(ftn, delegateType))
                VerificationError(VerifierError.DelegateCtor);
        }
        
        bool IsDelegateAssignable(MethodDesc ftn, TypeDesc delegateType)
        {
            var invokeMethod = delegateType.GetMethod("Invoke", null);
            if (invokeMethod == null)
                return false;

            var ftnSignature = ftn.Signature;
            var delegateSignature = invokeMethod.Signature;

            // Compare calling convention ignoring distinction between static and instance
            if ((ftnSignature.Flags & ~MethodSignatureFlags.Static) != (delegateSignature.Flags & ~MethodSignatureFlags.Static))
                return false;

            // Compare signature parameters
            if (ftnSignature.Length != delegateSignature.Length)
                return false;

            for (int i = 0; i < ftnSignature.Length; i++)
            {
                if (!IsAssignable(delegateSignature[i], ftnSignature[i]))
                    return false;
            }

            // Compare return type
            return IsAssignable(ftnSignature.ReturnType, delegateSignature.ReturnType);
        }
```

Reply:
Filed https://github.com/dotnet/runtime/issues/65010